tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc

```diff
@@ -3090,6 +3100,12 @@ ReductionCodegenInfo IrEmitterUnnested::ComputeReductionCodegenInfo(
   int64 num_threads_y = reduction_dimensions.is_row_reduction ? 1 : kWarpSize;
   int64 num_threads_x = [&] {
     if (reduction_dimensions.is_row_reduction) {
+      int cc_major = 0, cc_minor = 0;
+      ir_emitter_context_->device_description().cuda_compute_capability(
+          &cc_major, &cc_minor);
+      if (cc_major >= 6 && smallest_input_dtype_bits <= 16) {
+        return num_threads_x = kWarpSize * 4;
```

Comment:
Does this still make sense? I thought that with intra-block communication working it would make sense to saturate the block all the way to `kWarpSize * kWarpSize` if the dimension is big enough?

Reply:
In theory yes. My numbers also back this up. The exact reasons will depend of the exact shapes too. Here is some reasons:
- One SM can't use the full bandwidth per itself. So you need enough blocks to help saturate the IO. I did very small timing, but discarded those results as the noise was too high. There is 80 SM in V100, so ideally we would like at least 80 blocks.
- The last block stragler effect. If you have 81 blocks and 80 SM, then you have 1 that will cause the computation to be as long as if you had 160 blocks. More blocks mean smaller blocks, help mitigate this, as the smaller block will run faster.
- But as you just optimized, more blocks mean more output IO. So bigger block are better from that point of view.

So we need to find a balance in all that. In theory, we could also add a cost model to know if we have a IO or Compute bound kernel (fusion could create compute bound kernel). And in that case we could need different best configuration, as it could be register bound... I do not want to go there.

Reply:
Also, we need at least 64 threads per blocks if we want full occupancy. We are limited to 0.5 occupancy if we use 32 threads per blocks.